Comments fixes

mrcasals opened the pull request:
#### :tophat: What? Why?
This PR extracts the fixes related to the comments module in #5553, so that PR can be merged while we tweak some issues here.

#### :pushpin: Related Issues
- Related to #5553

#### :clipboard: Subtasks
- [x] Add `CHANGELOG` entry

### :camera: Screenshots (optional)
TBD

tramuntanal commented:
Hi @mrcasals this PR is in DRAFT and is labelled for v0.22, is there anything you need from our side? is it ready for review?

mrcasals commented:
Hi! This was extracted from #5553, which included a bunch of other changes. I moved some of these changes to this PR because they affected the UX (thanks @microstudi for the summary!). This PR was in draft in purpose because I had the intention of polishing it, but other priorities came and this fell off my list.

The idea behind this PR is to avoid loading all comments at once for a page with lots of comments. This scenario might cause an overload of the server, thus we added pagination.

The current PR behaves as it does because it has automatic pagination. This forces the comments form to be on top, because otherwise you'd need to load all comments before you're able to post a new one.

Possible solutions:

- Remove the automatic pagination in favor of a button to load newer comments and move the form back to the end of the page. The new comments should also be added at the end of the page. This *might* cause inconsistencies if a comment is created and a newer comments page is loaded
- Change the default order to newer comments first, leaving the form at the top. This is the easiest change considering the state of the PR, but I don't know how (or even if) this would affect participation /cc @decidim/product @carolromero  

Thoughts? @tramuntanal @microstudi @decidim/product @carolromero

carolromero commented:
> Possible solutions:
> - Remove the automatic pagination in favor of a button to load newer comments and move the form back to the end of the page. The new comments should also be added at the end of the page. This might cause inconsistencies if a comment is created and a newer comments page is loaded
> - Change the default order to newer comments first, leaving the form at the top. This is the easiest change considering the state of the PR, but I don't know how (or even if) this would affect participation /cc @decidim/product @carolromero

Hey everyone, my two cents on the options we have:

I'd go with the first solution @mrcasals suggests. I think that both in terms of usability and security it's the best option. I would also take care of the details, such as in the "Load more..." specifying the number of comments that is loaded each time (not sure if it has to be 10, 25 or 100).

As always, the ideal would be to do user tests, but in any case this would be the hypothesis I would test.

carolromero commented:
@mrcasals regarding the possible inconsistency that you comment on this option, I understand that it would be in the (rather unlikely?) case that it overlaps the loading of new comments with someone making a new one, right?

mrcasals commented:
@carolromero Imagine this scenario:

- 10 comments per page
- The resource has 21 comments (thus 3 pages). I see the 10 oldest comments and a button to load a new page
- I add a new comment. It gets added to the end of the page
	- It gets added **before** or **after** the "Load more" button"?
- I load more comments (2nd page).
	- What happens with the comment I created? Does it stick to the end? It disappears? New comments are added **after** my comment?
- I load the last page, where my comment should actually appear.
	- My comment (which was already loaded) disappears in favor of the newly loaded one? I see my comment 2 times?

microstudi commented:
@mrcasals  @carolromero 
Another approach is to do what Github does, which is to hide comments in the middle of the conversation, in this case the button "load more" loads comments starting at the oldest hidden comment and loads one "page" of comments (10). If there are still more comments to load, the button stays and allows the user to load more pages until the end.

This might also suffer from inconsistencies but might be solved by updating the last block of comments every time the user hits the "load more" button, so 2 pages are loaded, the first 10 hidden and the last 10 (where my comment should be). That last 10 will substitute the last block of comments.

In summary, with the 21 comments (3 pages):

1
2
3
4
5
6
7
8
9
10
...
LOAD MORE
...
21
<- new comments are posted here ->
FORM NEW COMMENT

What do you think, is that a possibility in terms of functionality and consistency?

carolromero commented:
@microstudi the approach you propose could also work well.

Yesterday I checked both Instagram and Reddit who also have this system, to see how they solve it.
Reddit for example has the form above and when you submit a new comment it is placed at the top of the list. Assuming there are 3 pages, if you load the second and third pages your comment still appears at the top. Only when you update the page does it reposition it in the order you have selected (newest, oldest, etc.)

It may be helpful to have a quick call to summarize the pros and cons of each option and make a decision.

tramuntanal commented:
@mrcasals @decidim/product this PR is scheduled for 0.23 but it is still in draft and seems stalled for a while. Is this planification correct?

andreslucena commented:
> @mrcasals @decidim/product this PR is scheduled for 0.23 but it is still in draft and seems stalled for a while. Is this planification correct? 

You're right, we'll remove the scheduled release for the moment.

carolromero commented:
I leave here this mockup for when we return to this issue
![imatge](https://user-images.githubusercontent.com/23509895/100260560-d9e95e00-2f49-11eb-8a9b-d4accb1c8703.png)

microstudi commented:
I should be much easier to implement this now that we don't have react

tramuntanal commented:
The "Load more" button is very small, I would make it as big as a comment bloc while keeping it distinguishable

andreslucena commented:
Hi! As this PR has stagnated, it's no longer valid (with the latest react removal on #6498) and the contract that this was being initially developed has also closed, I've moved the new definition and consensual solution of this problem to a new issue (see #7037) to further discuss and make a new PR. I'm closing this PR. 

Thanks everyone for giving your feedback!
